NEXT-12540 - [NEW] Plugins for Symfony Developers (#152)
shopware/docs: guides/plugins/plugins/plugins-for-symfony-developers.md

# Plugins for Symfony developers

## Overview

> This guide serves as an entry point for developers, who are familiar with the concepts of `Symfony bundles`.

Reviewer: ```suggestion
This guide serves as an entry point for developers familiar with the concepts of `Symfony bundles`.
```

## Prerequisites

This guide handles some base concepts of Shopware plugins. 
You may want to have a look at [plugin base guide](./plugin-base-guide.md) first.

As this guide also references the functionality of Symfony bundles,

> you should have at least basic knowledge of it.

Reviewer: ```suggestion
you should have at least a basic knowledge of it.
```

You may want to have a look or refresh your knowledge with help of Symfony's [bundle documentation](https://symfony.com/doc/current/bundles.html).

## Symfony bundles

A bundle is the Symfony's preferred way to provide additional third-party features to any Symfony application.
Those bundles are everywhere: Symfony even outsources many of its core features into external bundles.
The template engine `Twig`, the `Security` bundle, the `WebProfiler`, 
as well as many other third-party bundles can be installed on demand to extend your Symfony application in any way.
The Bundle System is Symfony's way of providing an extendable framework with plugin capabilities.

## Shopware plugins

> Shopware builds upon the `Symfony Bundle System` to extend its functionality even more.

Reviewer: ```suggestion
Shopware is building on the `Symfony Bundle System` to extend its functionality even more.
```

Not sure about that though, but it's the most common result if you search with deepl and google.

This allows the `Shopware Plugin System` to function as a traditional plugin system
with features like plugin lifecycles and more.

Whenever you create a Shopware plugin, you have to extend the `Shopware\Core\Framework\Plugin` class.

> If you investigate this class, you will see, that this class extends `Shopware\Core\Framework\Bundle`,

Reviewer: ```suggestion
If you investigate this class, you will see that this class extends `Shopware\Core\Framework\Bundle`,
```

> which in turn extends the Symfony's `Bundle` class:

Reviewer: ```suggestion
which in return extends the Symfony's `Bundle` class:
```

Am I right here? 🙈

{% code title="platform/src/Core/Framework/Bundle.php" %}
```php

> <?php declare(strict_types=1);

Reviewer 2: A while ago, I've built this slide to explain the inheritance hierarchy - maybe we can include that somehow as well?

![image](https://user-images.githubusercontent.com/7780750/111323910-3213af80-8662-11eb-9b7d-7825dda4b68c.png)

Reviewer 3: Nice! :-O

//...
use Symfony\Component\HttpKernel\Bundle\Bundle as SymfonyBundle;
//...

abstract class Bundle extends SymfonyBundle {
/** ... */
}
```

> 

Reviewer: ```suggestion
{% endcode %}

```

> As you can see, any Shopware plugin is internally a Symfony bundle as well, and will be handled as such by Symfony.

Reviewer: ```suggestion
As you can see, any Shopware plugin is also a Symfony bundle internally, and will be handled as such by Symfony.
```

> 

Reviewer 2: We could also add some plugin-specific additions to the bundle system (which are mentioned on the screenshot above) like the file-system, migrations.

Reviewer 3: Afaik symfony bundles also have a filesystem, migrations etc., don't they?
Might be wrong here

Author: I've added a small notice, that these extra functionality exists in Shopware plugins, but without explaining, what they really mean, as I think this would break the articles scope.

### Plugin lifecycle

> Shopware extends the `Symfony Bundle System` by some functionality to adjust its use for the Shopware ecosystem.

Reviewer: ```suggestion
Shopware extends the `Symfony Bundle System` with some functionality to adjust its use for the Shopware ecosystem.
```

> For you, as a plugin developer, the most important addition is the extended plugin lifecycle.

Reviewer: ```suggestion
For you as plugin developer, the most important addition is the extended plugin lifecycle.
```

A Shopware plugin runs through a lifecycle.

> Your plugin base class can implement the following methods to execute any sort of installation or maintenance tasks.  

Reviewer: ```suggestion
Your plugin's base class can implement the following methods to execute any sort of installation or maintenance tasks.  
```

| Lifecycle | Description |
| :--- | :--- |
| `install()` | Executed on plugin install |
| `postInstall()` | Executed **after** successful plugin install |
| `update` | Executed on plugin update |
| `postUpdate()` | Executed **after** successful plugin update |
| `uninstall()` | Executed on plugin uninstallation |
| `activate()` | Executed **before** plugin activation |

> | `deactivate()` | Executed **before** plugin deactivation |

Reviewer: At the end of the article, we usually write a small "Next step" section with some further guides which can be interesting for the reader. Would you like to add it, too?

```
{% endcode %}

As you can see, any Shopware plugin is also a Symfony bundle internally as well, and will be handled as such by Symfony.
A plugin adds support for some cases, specific to the Shopware environment.
These include, for example, handling plugin migrations and registering Shopware business events.

### Plugin lifecycle

As mentioned before, Shopware extends the `Symfony Bundle System` with some functionality to adjust its use for the Shopware ecosystem.
For you as plugin developer, the most important addition is the extended plugin lifecycle.

A Shopware plugin runs through a lifecycle.
Your plugin's base class can implement the following methods to execute any sort of installation or maintenance tasks.  

| Lifecycle | Description |
| :--- | :--- |
| `install()` | Executed on plugin install |
| `postInstall()` | Executed **after** successful plugin install |
| `update` | Executed on plugin update |
| `postUpdate()` | Executed **after** successful plugin update |
| `uninstall()` | Executed on plugin uninstallation |
| `activate()` | Executed **before** plugin activation |
| `deactivate()` | Executed **before** plugin deactivation |

## Next steps

> * []()

Author: Any Ideas for some next steps ?

Reviewer 3: Guess just the plugin base guide. Like "Now you know the differences - time to get started" or smth

Author: Did think about this aswell, but the plugin base guide is already in listed as a prerequisite

Author: I added some links to symfony-specific topics

# Plugins for Symfony developers

## Overview

This guide serves as an entry point for developers familiar with the concepts of `Symfony bundles`.

## Prerequisites

This guide handles some base concepts of Shopware plugins. 

> You may want to have a look at [plugin base guide](./plugin-base-guide.md) first.

Reviewer: ```suggestion
Therefore, you may want to have a look at [plugin base guide](./plugin-base-guide.md) first.
```

As this guide also references the functionality of Symfony bundles,
you should have at least a basic knowledge of it.
You may want to have a look or refresh your knowledge with help of Symfony's [bundle documentation](https://symfony.com/doc/current/bundles.html).

## Symfony bundles

A bundle is the Symfony's preferred way to provide additional third-party features to any Symfony application.
Those bundles are everywhere: Symfony even outsources many of its core features into external bundles.
The template engine `Twig`, the `Security` bundle, the `WebProfiler`, 
as well as many other third-party bundles can be installed on demand to extend your Symfony application in any way.
The Bundle System is Symfony's way of providing an extendable framework with plugin capabilities.

## Shopware plugins

Shopware is building upon the `Symfony Bundle System` to extend its functionality even more.

> This allows the `Shopware Plugin System` to function as a traditional plugin system

Reviewer: ```suggestion
This allows the Shopware Plugin System to function as a traditional plugin system
```

with features like plugin lifecycles and more.

Whenever you create a Shopware plugin, you have to extend the `Shopware\Core\Framework\Plugin` class.
If you investigate this class, you will see that this class extends `Shopware\Core\Framework\Bundle`,
which in return extends the Symfony's `Bundle` class:

> {% code %}

Reviewer: Are you able to give an example path?

Author: Nope, this is an abstract example.
